rbd/migration.go

```diff
@@ -0,0 +1,240 @@
+		migrationOpts.destImageName,
+		C.rbd_image_options_t(rio.options))
+
+	return getError(ret)
+}
+
+// MigrationExecute executes a migration starting
+// the full migration of data into image from source
+//
+// Implements:
+//  int rbd_migration_execute(rados_ioctx_t ioctx,
+// 							  const char *image_name);
+//
+func MigrationExecute(ioctx *rados.IOContext, name string) error {
+	cName := (*C.char)(C.CBytes([]byte(name)))
+
+	defer func() {
+		C.free(unsafe.Pointer(cName))
+	}()
+
+	ret := C.rbd_migration_execute(
+		cephIoctx(ioctx),
+		cName)
+	return getError(ret)
+}
+
+// MigrationCommit commits a migration after execution
+// breaking the relationship of image to the source
+//
+// Implements:
+//  int rbd_migration_commit(rados_ioctx_t ioctx,
+// 							 const char *image_name);
+//
+func MigrationCommit(ioctx *rados.IOContext, name string) error {
+	cName := (*C.char)(C.CBytes([]byte(name)))
+
+	defer func() {
+		C.free(unsafe.Pointer(cName))
+	}()
+
+	ret := C.rbd_migration_commit(
+		cephIoctx(ioctx),
+		cName)
+	return getError(ret)
+}
+
+// MigrationAbort aborts a migration in progress
+// breaking the relationship of image to the source
+//
+// Implements:
+//  int rbd_migration_abort(rados_ioctx_t ioctx,
+// 							const char *image_name);
+//
+func MigrationAbort(ioctx *rados.IOContext, name string) error {
+	cName := (*C.char)(C.CBytes([]byte(name)))
+
+	defer func() {
+		C.free(unsafe.Pointer(cName))
+	}()
+
+	ret := C.rbd_migration_abort(
+		cephIoctx(ioctx),
+		cName)
+	return getError(ret)
+}
+
+// MigrationImageState denotes the current migration status of a given image
+type MigrationImageState int
+
+const (
+	// MigrationImageUnknown is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_UNKNOWN from librbd.
+	MigrationImageUnknown = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_UNKNOWN)
+	// MigrationImageError is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_ERROR from librbd.
+	MigrationImageError = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_ERROR)
+	// MigrationImagePreparing is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_PREPARING from librbd.
+	MigrationImagePreparing = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_PREPARING)
+	// MigrationImagePrepared is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_PREPARED from librbd.
+	MigrationImagePrepared = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_PREPARED)
+	// MigrationImageExecuting is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_EXECUTING from librbd.
+	MigrationImageExecuting = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_EXECUTING)
+	// MigrationImageExecuted is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_EXECUTED from librbd.
+	MigrationImageExecuted = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_EXECUTED)
+	// MigrationImageAborting is the representation of
+	// RBD_IMAGE_MIGRATION_STATE_ABORTING from librbd.
+	MigrationImageAborting = MigrationImageState(C.RBD_IMAGE_MIGRATION_STATE_ABORTING)
+)
```

Comment:
Style nitpick: I generally like to see constants defined near the top of the file, before we start defining structures (unless there are many structures in the with with many different constants in the same file).

Reply:
Done